### ISSUE TITLE
Testing with typescript + jest + ts-jest

### ISSUE THREAD

**Issue by runia1** [zeit/next.js#8663]
## To Reproduce

- `npm i -D jest typescript ts-jest @types/jest react-test-renderer`
- add a `jest.config.js` with following content:
```
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node'
};
```
- add a `tsconfig.json` with the following content:
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve"
  },
  "exclude": [
    "node_modules"
  ],
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ]
}
```
**NOTE: this is the default generated by next.js when it determines you're using typescript**

- add a test file `__tests__/index.test.tsx` with the following contents:
```
import React from 'react';
import TestRenderer from 'react-test-renderer';
import Index from '../pages/index';

describe('Index', () => {
    it('should render without throwing an error', () => {
        const testRenderer = TestRenderer.create(<Index />);
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });
});
```
- add a test script to `package.json` --> "test": "jest"
- try to run tests `npm run test`

At this point you should see the following error: 
```
 ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /Users/mar/NodeJSProjects/lawn-box/__tests__/index.test.tsx:11
            var testRenderer = react_test_renderer_1.default.create(<index_1.default />);
                                                                    ^

    SyntaxError: Unexpected token <
```

## Expected behavior

You expect the test to be transformed and pass.

## How to solve

This issue is caused because jsx is not being transformed and jest doesn't know how to parse it.  The fix is very simple, in your `tsconfig.json` you need to set the "jsx" key to "react" rather than "preserve".  Unfortunately next's build expects it to be "preserve" and automatically changes it back in `tsconfig.json` :(  The way I got around this was to create a seperate `tsconfig.jest.json`.  Here are the steps:

- create a file `tsconfig.jest.json` with the following contents:
```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "jsx": "react"
  }
}
```
NOTE: here we are extending the base tsconfig so most of our settings are getting applied, and simply overriding the "jsx" key.

- configure `jest.config.js` to use our new `tsconfig.jest.json`:
```
module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  globals: {
    // we must specify a custom tsconfig for tests because we need the typescript transform
    // to transform jsx into js rather than leaving it jsx such as the next build requires.  you
    // can see this setting in tsconfig.jest.json -> "jsx": "react"
    "ts-jest": {
      tsConfig: "tsconfig.jest.json"
    }
  }
};
```
- run tests --> `npm run test`
- congrats you got next.js + typescript + jest + ts-jest working with jsx 🎉🎉🎉

Note to the maintainers: Feel free to close this issue, hoping it will help others who may run into this.

**Comment by Janpot** [comment]
Maybe it should be (part of) one of the examples?

**Comment by mifrej** [comment]
Yes it should be documented. I spend some time to figure that out myself

**Comment by codemilli** [comment]
Can we assure that the result is always same?
if jsx field is `preserve` as you know, it passes the role of parsing jsx to next parser weather typescript parse the jsx when jsx field is `react`.

**Comment by runia1** [comment]
@codemilli I'm guessing here but I think next's parser is just wrapping the typescript parser when it finds that you have a tsconfig.json.  So if you set up ts-jest to use the same config you should be alright.

**Comment by timneutkens** [comment]
> I'm guessing here but I think next's parser is just wrapping the typescript compiler when it finds that you have a tsconfig.json

We use @babel/preset-typescript to compile the code and don't actually need a tsconfig.json, however for consistency and integration with code editors like vscode we force the tsconfig to have the exact same settings as the way that we compile. We also do type checking in parallel which does use `tsc` but that is only for type checking and not compiling the code.

**Comment by timneutkens** [comment]
I also don't think you need ts-jest in particular if babel is ran against the test code correctly

**Comment by runia1** [comment]
@codemilli there's your answer as to how nextjs is compiling typescript ^.  They're using babel with a typescript preset vs. `tsc` so you're not likely going to get the exact same result.  I would hope that `@babel/preset-typescript` is transpiling to near the same output however.

As far as not using ts-jest, we would need some way to tell jest to plug in to the babel config then, and transpile with babel prior to attempting to run tests.  Is there any documentation for how to do that?

**Comment by dballance** [comment]
You should be able to simply add a `.babelrc` like [this example](https://github.com/zeit/next.js/blob/canary/examples/with-jest-typescript/.babelrc). 

If you add the `.babelrc` and `jest.config.js` and `jest.setup.js` like the example, you simply need `jest`, `@types/jest` and then renderer of your choice (`enzyme` or `react-test-renderer`).

Running `jest` just works at that point.

**Comment by Timer** [comment]
We just landed a new Jest example.

**Comment by CapySloth** [comment]
> We just landed a new Jest example.

Link to pr?

**Comment by Timer** [comment]
https://github.com/zeit/next.js/tree/canary/examples/with-jest

**Comment by Prottoy2938** [comment]
I think Jest Test Environment would be `jsdom` instead of `node`.
 

>  testEnvironment: "jsdom",

**Comment by mehmetnyarar** [comment]
I encounter the following error if I don't import React in both `page.tsx` and `page.spec.tsx`:

> 'React' refers to a UMD global, but the current file is a module. Consider adding an import instead.

Any ideas?

**Comment by Prottoy2938** [comment]
@mehmetnyarar , I'm not 100% sure if that problem is related to this issue, but if you're having trouble using **Nextjs** with **Jest**, you can try out the boilerplate that I've made: https://github.com/Prottoy2938/nextjs-typescript-eslint-jest-template.git

Its a boilerplate where it has the basic configuration for **Nextjs** with **TypeScript**, **ESLint** and **Jest**

**Comment by jf-t** [comment]
@mehmetnyarar I am getting the same error as you did. Did you find a workaround? I have a year-old Next.js app and have never specifically imported React but it seems with jest I have to do it... Seems like there must be a way around this.

**Comment by MatheusFrez** [comment]
Thanks a lot @runia1 tsconfig.jest.json works like a charm!! 🤓🤓

**Comment by majew7** [comment]
I followed @runia1 steps, but oddly it did not solve the issue to get Jest and TypesScript working in NextJS:


```
    /Users/person1/Documents/code/Git/arg-matrix/pages/index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){import Head from 'next/head'
                                                                                             ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      1 | import React from 'react';
      2 | import TestRenderer from 'react-test-renderer';
    > 3 | import Index from '../pages/index';
        | ^
      4 |
      5 | describe('Index', () => {
      6 |     it('should render without throwing an error', () => {

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)
      at Object.<anonymous> (__tests__/index.test.tsx:3:1)

Test Suites: 1 failed, 1 total
```

**Comment by Vijayan-Murugan-Mvp** [comment]
({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){.App {
                                                                                             ^

    SyntaxError: Unexpected token '.'

      1 | import * as React from 'react'
    > 2 | import './App.css';
        | ^
      3 | import { ToastContainer } from 'react-toastify';
      4 |
      5 | function App() {

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1350:14)
      at Object.<anonymous> (src/App.tsx:2:1)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        3.85 s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! alpha-ui@0.1.0 test: `jest`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the alpha-ui@0.1.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

can i get help i have this error.

**Comment by pavlovoloshchukbeetroot** [comment]
Why something considering so simple yet too complicated to establish -> Jest + Ts + NextJs. None of the available receipts just don't work. Already spent 2 days specifically on jest.

**Comment by chrisgibbs44** [comment]
I was getting the React UMD @mehmetnyarar , I fixed it by putting this in `jest.config.js`:
```
  transform: {
    "^.+\\.(ts|tsx)$": "<rootDir>/node_modules/babel-jest",
```
(before it was this: `'^.+\\.(ts)$': 'ts-jest',`)

**NB you may also need to add `js|jsx` to the above transform and install babel-jest** - in our case we're just testing `.tsx` files and we already had babel-jest installed, so we created a `.babelrc` just with this as its contents:
```
{
  "presets": ["next/babel"]
}
```
and now jest works with .tsx files! 🚀 

(thanks to https://dev.to/ericdouglas/how-to-configure-jest-on-a-next-js-project-2ic6 for the `.babelrc` fix)

**Comment by mareksuscak** [comment]
None of the solutions are working as expected. When `babel-jest` is used, path aliases defined in `tsconfig.json` are not working. When `ts-jest` is used, JSX is not being compiled correctly despite creating a separate `tsconfig.test.json` and linking it via `globals`. Has anyone actually solved this issue?

**Comment by adamIncep** [comment]
Can confirm that this is still an issue. I was able to fix most of the problems with: 

Adding ts-jest lib, change the testEnvironment to 'jsdom' here:  

 jest.config.js: 
 `module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
};`

tsconfig.json:

Re-added the libs, allowJS: true, jsx: "react-jsx"

`{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "react-jsx"
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ],
  "exclude": [
    "node_modules"
  ]
}
`

But why you guys changing back tsconfig.json?? to "jsx" : "preserve" ????

With preserve the tests do not run properly with the above config.

**Comment by chrisgibbs44** [comment]
@mareksuscak yeah good point about the path aliases not working, probably not a great long term solution

after further investigation with the `tsconfig.test.json` method outlined in the original post, it seems to be this extra fix that makes it work:

```{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "jsx": "react-jsx"
  }
}
```
--> when "jsx" is set to "react" it doesn't work (React UMD error), whereas "react-jsx" does work for me 👍

**Comment by mareksuscak** [comment]
@chrisgibbs44 yeah, 2 hours later I figured out the same thing. Also, for anyone else who's been wondering, here's my `jest.config.ts`

```
import type { Config } from '@jest/types'

import { pathsToModuleNameMapper } from 'ts-jest/utils'
import { compilerOptions } from './tsconfig.json'

const config: Config.InitialOptions = {
  roots: ['<rootDir>'],
  // Node modules are referenced twice on purpose, this fixes smoe issue that I will link
  // if I'm able to find it.
  moduleDirectories: ['node_modules', '<rootDir>/node_modules', '.'],
  moduleFileExtensions: ['js', 'ts', 'tsx', 'json'],
  testPathIgnorePatterns: ['<rootDir>[/\\\\](node_modules|.next)[/\\\\]'],
  transformIgnorePatterns: ['[/\\\\]node_modules[/\\\\].+\\.(ts|tsx)$'],
  collectCoverageFrom: ['**/*.{ts,tsx}', '!**/*.d.ts', '!**/node_modules/**'],
  watchPlugins: ['jest-watch-typeahead/filename', 'jest-watch-typeahead/testname'],
  moduleNameMapper: pathsToModuleNameMapper({
    ...compilerOptions.paths,
    '\\.(css|scss)$': ['identity-obj-proxy'],
    '\\.(jpg|jpeg|png|gif|svg)$': ['<rootDir>/test/__mocks__/fileMock.js'],
  }),
  testURL: `http://localhost`,
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
  // There are other presets to choose from depending on whether your
  // project uses both ts and js files and how you want to deal with them.
  preset: 'ts-jest',
  globals: {
    'ts-jest': {
      // The contents of this file should be what @chrisgibbs44 shared above
      tsconfig: 'tsconfig.test.json',
      isolatedModules: true,
    },
  },
}

export default config
```

**Comment by DerekHill** [comment]
Note that the `tsConfig` jest config option as removed in ts-jest 27, in favour of `tsconfig`.  So in your `jest.config.js`, instead of:

```javascript
module.exports = {
  preset: "ts-jest",
  globals: {
    "ts-jest": {
      tsConfig: "tsconfig.test.json",
    },
  },
};
```

Be sure to do:

```javascript
module.exports = {
  preset: "ts-jest",
  globals: {
    "ts-jest": {
      tsconfig: "tsconfig.test.json",
    },
  },
};
```

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
